addons/binding/org.openhab.binding.chamberlainmyq/src/main/java/org/openhab/binding/chamberlainmyq/internal/discovery/ChamberlainMyQDeviceDiscoveryService.java

```diff
@@ -0,0 +1,145 @@
+import java.util.Map;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.chamberlainmyq.config.ChamberlainMyQDeviceConfig;
+import org.openhab.binding.chamberlainmyq.handler.ChamberlainMyQGatewayHandler;
+import org.openhab.binding.chamberlainmyq.handler.ChamberlainMyQGatewayHandler.RequestCallback;
+import org.openhab.binding.chamberlainmyq.internal.ChamberlainMyQHandlerFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+/**
+ * Send DiscoveryService Callbacks
+ *
+ * @author Scott Hanson - Initial contribution
+ */
+
+public class ChamberlainMyQDeviceDiscoveryService extends AbstractDiscoveryService {
+    private final Logger logger = LoggerFactory.getLogger(ChamberlainMyQDeviceDiscoveryService.class);
+    private ChamberlainMyQGatewayHandler hubHandler;
+
+    public ChamberlainMyQDeviceDiscoveryService(ChamberlainMyQGatewayHandler hubHandler)
+            throws IllegalArgumentException {
+        super(ChamberlainMyQHandlerFactory.DISCOVERABLE_DEVICE_TYPES_UIDS, 10);
+
+        this.hubHandler = hubHandler;
+    }
+
+    private ScheduledFuture<?> scanTask;
+
+    @Override
+    protected void startScan() {
+        if (this.scanTask == null || this.scanTask.isDone()) {
+            this.scanTask = scheduler.schedule(new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        readDeviceDatabase();
+                    } catch (Exception e) {
+                        logger.error("Error scanning for devices", e);
+
+                        if (scanListener != null) {
+                            scanListener.onErrorOccurred(e);
+                        }
+                    }
+                }
+            }, 0, TimeUnit.SECONDS);
+        }
+    }
+
+    protected void addMyQDevice(ThingTypeUID thinkType, ChamberlainMyQDeviceConfig config) {
+        logger.debug("New Device {}", config.toString());
+
+        ThingUID hubUID = this.hubHandler.getThing().getUID();
+        ThingUID uid = new ThingUID(thinkType, hubUID, config.getDeviceId());
+
+        Map<String, Object> properties = config.getProperties();
+        DiscoveryResult result = DiscoveryResultBuilder.create(uid).withLabel(config.getDescription())
+                .withProperties(properties).withBridge(hubUID).build();
+
+        thingDiscovered(result);
+
+        logger.debug("Discovered {}", uid);
+    }
+
+    protected void enumerateDevices(JsonObject gatewayResponse) {
+        JsonElement deviceData = gatewayResponse.get("Devices");
```

Comment:
You are parsing the json library imperative. Why do you not use gsons java object mapping?
This lowers maintenance effort considerably and shortens the code.

Reply:
updated to gson